engine/runtime-instrument-common/src/main/java/org/enso/interpreter/instrument/command/SetExecutionEnvironmentCommand.java

```diff
@@ -41,10 +41,12 @@ public Future<BoxedUnit> executeAsynchronously(RuntimeContext ctx, ExecutionCont
   private void setExecutionEnvironment(
       Runtime$Api$ExecutionEnvironment executionEnvironment, UUID contextId, RuntimeContext ctx) {
     var logger = ctx.executionService().getLogger();
-    var contextLockTimestamp = ctx.locking().acquireContextLock(contextId);
-    var writeLockTimestamp = ctx.locking().acquireWriteCompilationLock();
+    long contextLockTimestamp = 0;
+    long writeLockTimestamp = 0;
 
     try {
+      contextLockTimestamp = ctx.locking().acquireContextLock(contextId);
```

Comment:
I see some inconsistency. Here we acquire the lock in `try` block. While in the [InvalidateModuleIndex...](https://github.com/enso-org/enso/pull/8270/files#r1389480949) we acquire the lock before entering the `try` block.

The inconsistency shall be unified.

However doing it with `try/catch` is tedious. Wouldn't it be more functional to have `ctx.locking().withContextLock(Runnable run)`? Then the consistency would be encapsulated inside of such `withContextLock` method.

Reply:
Btw. every project complex enough needs its own "locking class", right?

Here is my [Mutex](https://bits.netbeans.org/dev/javadoc/org-openide-util/org/openide/util/Mutex.html) written at beginnings of NetBeans. It uses the `Runnable` callbacks approach as it (in my opinion) better reflects the structured nature of locking.

However over time we also had to offer [non structured variant](https://bits.netbeans.org/dev/javadoc/org-openide-util/org/openide/util/Mutex.Privileged.html). We didn't want some random users of NetBeans API to take a lock and forget to unlock it - as such the `Priviledged` API is only available to _creator_ of the `Mutex` - regular users have to use the `Runnable` variants.

Reply:
`withXYZ` sounds like a good idea (I thought you didn't like using `Runnable`?)

Reply:
`Runnable` is fine. Lambdas have some overhead, but the overhead of forgetting to unlock is probably higher.